rpc/personal_api.go

```diff
@@ -1,30 +1,168 @@
+	"github.com/ethereum/go-ethereum/crypto"
 )
 
 // PersonalEthAPI is the eth_ prefixed set of APIs in the Web3 JSON-RPC spec.
 type PersonalEthAPI struct {
-	cliCtx    sdkcontext.CLIContext
-	nonceLock *AddrLocker
+	cliCtx      sdkcontext.CLIContext
+	ethAPI      *PublicEthAPI
+	nonceLock   *AddrLocker
+	keys        []emintcrypto.PrivKeySecp256k1
+	keybaseLock sync.Mutex
 }
 
 // NewPersonalEthAPI creates an instance of the public ETH Web3 API.
-func NewPersonalEthAPI(cliCtx sdkcontext.CLIContext, nonceLock *AddrLocker) *PersonalEthAPI {
+func NewPersonalEthAPI(cliCtx sdkcontext.CLIContext, ethAPI *PublicEthAPI, nonceLock *AddrLocker, keys []emintcrypto.PrivKeySecp256k1) *PersonalEthAPI {
 	return &PersonalEthAPI{
 		cliCtx:    cliCtx,
+		ethAPI:    ethAPI,
 		nonceLock: nonceLock,
+		keys:      keys,
 	}
 }
 
+func (e *PersonalEthAPI) getKeybaseInfo() ([]keyring.Info, error) {
+	e.keybaseLock.Lock()
+	defer e.keybaseLock.Unlock()
+
+	if e.cliCtx.Keybase == nil {
+		keybase, err := keyring.NewKeyring(
+			sdk.KeyringServiceName(),
+			viper.GetString(flags.FlagKeyringBackend),
+			viper.GetString(flags.FlagHome),
+			e.cliCtx.Input,
+			emintcrypto.EthSecp256k1Options()...,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		e.cliCtx.Keybase = keybase
+	}
+
+	return e.cliCtx.Keybase.List()
+}
+
+// ImportRawKey stores the given hex encoded ECDSA key into the key directory,
+// encrypting it with the passphrase.
+func (e *PersonalEthAPI) ImportRawKey(privkey, password string) (common.Address, error) {
+	return common.Address{}, nil
+}
+
+// ListAccounts will return a list of addresses for accounts this node manages.
+func (e *PersonalEthAPI) ListAccounts() ([]common.Address, error) {
+	addrs := []common.Address{}
+	infos, err := e.getKeybaseInfo()
+	if err != nil {
+		return addrs, err
+	}
+
+	for _, info := range infos {
+		addressBytes := info.GetPubKey().Address().Bytes()
+		addrs = append(addrs, common.BytesToAddress(addressBytes))
+	}
+
+	return addrs, nil
+}
+
+// LockAccount will lock the account associated with the given address when it's unlocked.
+func (e *PersonalEthAPI) LockAccount(address common.Address) bool {
+	for i, key := range e.keys {
+		if !bytes.Equal(key.PubKey().Address().Bytes(), address.Bytes()) {
+			continue
+		}
+
+		tmp := make([]emintcrypto.PrivKeySecp256k1, len(e.keys)-1)
+		copy(tmp[:i], e.keys[:i])
+		copy(tmp[i:], e.keys[i+1:])
+		e.keys = tmp
+		return true
+	}
+
+	return false
+}
+
+// NewAccount will create a new account and returns the address for the new account.
+func (e *PersonalEthAPI) NewAccount(password string) (common.Address, error) {
+	_, err := e.getKeybaseInfo()
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	info, _, err := e.cliCtx.Keybase.CreateMnemonic("key_"+time.Now().Format(time.RFC3339), keyring.English, password, emintcrypto.EthSecp256k1)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	return common.BytesToAddress(info.GetPubKey().Address().Bytes()), nil
```

Comment:
nit, we should add the same logs:

https://github.com/ethereum/go-ethereum/blob/04926db204232b7103b8eb14bf3f8e9de129ad44/internal/ethapi/api.go#L278-L291

Reply:
done